controller/scheduler-v2/scheduler.go

```diff
@@ -0,0 +1,836 @@
+			err = s.SyncFormations()
+			continue
+		case <-s.syncJobs:
+			err = s.SyncJobs()
+			continue
+		default:
+		}
+
+		// Finally, handle triggering cluster changes
+		select {
+		case <-s.rectifyJobs:
+			err = s.RectifyJobs()
+		case <-time.After(50 * time.Millisecond):
+			// block so that
+			//	1) mutate events are given a chance to happen
+			//  2) we don't spin hot
+		}
+	}
+	return nil
+}
+
+func (s *Scheduler) SyncJobs() (err error) {
+	fLog := s.log.New("fn", "SyncJobs")
+
+	defer func() {
+		s.sendEvent(NewEvent(EventTypeClusterSync, err, nil))
+	}()
+
+	newSchedulerJobs := make(map[string]*Job)
+	hosts, err := s.getHosts()
+	if err != nil {
+		return errors.New("Unable to query hosts")
+	}
+	for _, h := range hosts {
+		hLog := fLog.New("host.id", h.ID())
+		hLog.Info("getting jobs list")
+		var activeJobs map[string]host.ActiveJob
+		activeJobs, err = h.ListJobs()
+		if err != nil {
+			hLog.Error("error getting jobs list", "err", err)
+			continue
+		}
+		hLog.Info("active jobs", "count", len(activeJobs))
+		for _, activeJob := range activeJobs {
+			s.handleActiveJob(&activeJob)
+			if j, ok := s.jobs[activeJob.Job.ID]; ok {
+				newSchedulerJobs[j.JobID] = s.jobs[j.JobID]
+			}
+		}
+	}
+
+	s.jobs = newSchedulerJobs
+	triggerChan(s.rectifyJobs)
+
+	return err
+}
+
+func (s *Scheduler) SyncFormations() (err error) {
+	log := s.log.New("fn", "SyncFormations")
+
+	defer func() {
+		s.sendEvent(NewEvent(EventTypeFormationSync, err, nil))
+	}()
+
+	log.Info("getting formations")
+	apps, err := s.AppList()
+	if err != nil {
+		return err
+	}
+	for _, app := range apps {
+		fs, err := s.FormationList(app.ID)
+		if err != nil {
+			return err
+		}
+		for _, f := range fs {
+			_, err = s.updateFormation(f)
+		}
+	}
+	triggerChan(s.rectifyJobs)
+	return err
+}
+
+func (s *Scheduler) RectifyJobs() (err error) {
+	if !s.isLeader {
+		return errors.New("RectifyJobs(): this scheduler is not the leader")
+	}
+
+	log := s.log.New("fn", "RectifyJobs")
+	defer func() {
+		if err != nil {
+			log.Error("error rectifying jobs", "err", err)
+		}
+		s.sendEvent(NewEvent(EventTypeRectifyJobs, err, nil))
+	}()
+
+	log.Info("Existing jobs", "jobs", s.jobs)
+	fj := NewPendingJobs(s.jobs, MergePendingJobs(s.pendingStarts, s.pendingStops))
+
+	for fKey := range fj {
+		schedulerFormation, ok := s.formations[fKey]
+
+		if !ok {
+			cf, err := s.GetFormation(fKey.AppID, fKey.ReleaseID)
+			if err != nil {
+				log.Error("Job exists without formation")
+				continue
+			}
+			schedulerFormation, err = s.updateFormation(cf)
+		}
+
+		schedulerProcs := schedulerFormation.Processes
+		clusterProcs := fj.GetProcesses(fKey)
+
+		if eq := reflect.DeepEqual(clusterProcs, schedulerProcs); !eq {
+			log.Info("Updating processes", "formation.processes", schedulerProcs, "cluster.processes", clusterProcs)
+			schedulerFormation.Processes = clusterProcs
+
+			diff := schedulerFormation.Update(schedulerProcs)
+			s.sendDiffRequests(schedulerFormation, diff)
+		}
```

Comment:
This needs to ignore jobs with no type (otherwise known as "one-off jobs", like what `flynn run bash` starts) as they are currently being killed shortly after starting.